Initial commits for review

tssala23 opened the pull request:
PR for reviewing work on issue https://github.com/nerc-project/operations/issues/333

larsks commented on README.md:
```diff
@@ -0,0 +1,11 @@
+# Nerc-Cluster-Templating
+This repo has an Anisble playbook to create overlay dir to be applied on NERC for new clusters.
```
```suggestion
This repo has an Ansible playbook to create an overlay dir for a standard NERC cluster.
```

larsks commented on README.md:
```diff
@@ -0,0 +1,11 @@
+# Nerc-Cluster-Templating
+This repo has an Anisble playbook to create overlay dir to be applied on NERC for new clusters.
+
+There a few variable to be set in the [vars.yaml](https://github.com/tssala23/nerc-cluster-templating/blob/main/roles/cluster_template/vars/vars.yaml) file.
```
```suggestion
There a few variable to be set in the [vars.yaml](roles/cluster_template/vars/vars.yaml) file.
```

larsks commented on README.md:
```diff
@@ -0,0 +1,11 @@
+There are two example outputs in the outputs directory.
+
+## Usage
+After setting the variables in the [vars.yaml](https://github.com/tssala23/nerc-cluster-templating/blob/main/roles/cluster_template/vars/vars.yaml) file, run the playbook `ansible-playbook playbook.yaml`.
```
```suggestion
After setting the variables in the [vars.yaml](roles/cluster_template/vars/vars.yaml) file, run the playbook `ansible-playbook playbook.yaml`.
```

larsks commented on README.md:
```diff
@@ -0,0 +1,11 @@
+## Usage
+After setting the variables in the [vars.yaml](https://github.com/tssala23/nerc-cluster-templating/blob/main/roles/cluster_template/vars/vars.yaml) file, run the playbook `ansible-playbook playbook.yaml`.
+
+The resulting cluster dir, which would be transferred to the [overlays dir](https://github.com/OCP-on-NERC/nerc-ocp-config/tree/main/cluster-scope/overlays) is outputted to the outputs dir.
```
```suggestion
The resulting cluster dir, which would be transferred to the [cluster-scope overlays dir](https://github.com/OCP-on-NERC/nerc-ocp-config/tree/main/cluster-scope/overlays) is outputted to the outputs dir.
```

larsks commented on output/external-example/certificates/default_api_certificate.yaml:
I'm not sure you want to keep this example output as part of the repository. Unless you add some tooling to automatically re-generate the examples, it's very easy for them to fall out of sync with respect to your playbooks.

larsks commented on playbook.yaml:
```diff
@@ -0,0 +1,8 @@
+  connection: local
+  gather_facts: false
+  vars_files:
+    - roles/cluster_template/vars/vars.yaml
+    - roles/cluster_template/vars/constants.yaml
```
You should not be including vars files from inside your roles. If you configure `roles/<role>/vars/main.yaml` or `roles/<role>/defaults/main.yaml` they will be discovered automatically by the role. In general prefer `defaults/main.yaml` over `vars/main.yaml` because the values in `defaults` are designed to be [easy to override](https://docs.ansible.com/ansible/2.9/user_guide/playbooks_reuse_roles.html#role-default-variables).

If you need vars files from outside the role, put them somewhere else.

Ansible will load `group_vars/all.yaml` automatically (or `group_vars/all/<whatever>.yaml` if you would like more than one file).

larsks commented on roles/cluster_template/tasks/main.yaml:
```diff
@@ -0,0 +1,176 @@
+    state: directory
+    mode: '0755'
+  loop:
+    - "output/{{ clusterName }}"
```
Here (and everywhere else), `output` should probably be a variable (so that it's possible to place the generated output wherever we want, including in the appropriate cluster overlay directory).

larsks commented on roles/cluster_template/tasks/main.yaml:
```diff
@@ -0,0 +1,176 @@
+    - "output/{{ clusterName }}/externalsecrets"
+    - "output/{{ clusterName }}/nodenetworkconfigurationpolicies"
+
+#Create dirs only found in external cluster
```
"external" may be the wrong word here. We have effectively two axes we can use to describe clusters:

- `external`/`internal` -- this describes how the cluster is accessed. "internal" clusters require VPN access; "external" clusters have a public ip address.

- `user`/`admin` -- this describes how we expect the cluster to be *used*. On a "user" cluster we permit non-admin access and people can run arbitrary workloads in their namespaces. An "admin" cluster only allows logins from the NERC team.

So e.g. the observability cluster is an "external admin" cluster, while the test cluster is effectively an "internal user" cluster.

larsks commented on roles/cluster_template/templates/certificates/default-api-certificate.j2:
```diff
@@ -0,0 +1,14 @@
+    name: letsencrypt-production-dns01
+    kind: Issuer
+  secretName: default-api-certificate
+  duration: 2160h0m0s
+  renewBefore: 360h0m0s
```
These values should be variables (that default to the existing values).

larsks commented on roles/cluster_template/templates/certificates/default-api-certificate.j2:
```diff
@@ -0,0 +1,14 @@
+  duration: 2160h0m0s
+  renewBefore: 360h0m0s
+  dnsNames:
+    - "api.shift.nerc.mghpcc.org"
```
This needs to be a variable (because it's the api hostname for the cluster).

larsks commented on roles/cluster_template/templates/certificates/default-ingress-certificate.j2:
```diff
@@ -0,0 +1,14 @@
+  duration: 2160h0m0s
+  renewBefore: 360h0m0s
+  dnsNames:
+    - "*.apps.shift.nerc.mghpcc.org"
```
This needs to be a a variable.

larsks commented on roles/cluster_template/tasks/main.yaml:
```diff
@@ -0,0 +1,176 @@
+  when: clusterType == "internal"
+
+#Overlay Kustomization.yaml file
+- template:
```
You can probably replace most of these `template` paths with a loop. If you structure the `templates` directory appropriately, then the output path is the same relative path as the template path.

Consider the following example; first we iterate over directories and create those in the target; then we iterate over template files and render them in the appropriate destination:

```
- name: create directories
  file:
    state: directory
    path:  "{{ dest }}"
  vars:
    dest: '{{ "output/%s/%s" % (clusterName, item.path) }}'
  when: >-
    item.state == "directory"
  loop: "{{ query('filetree', 'templates/')}}"
  loop_control:
    label: "{{ item.path }}"

- name: apply templates
  template:
    src:  "{{ src }}"
    dest:  "{{ dest }}"
  vars:
    src: "{{ item.path }}"
    dest: '{{ "output/%s/%s" % (clusterName, item.path) }}'
  when: >-
    item.state == "file"
  loop: "{{ query('filetree', 'templates/')}}"
  loop_control:
    label: "{{ item.path }}"
```

The above won't work as written (the existing structure of the templates directory isn't appropriate), but maybe it gives you some ideas.

This is particularly nice because in order to introduce a new template, you simply have to create it in the appropriate location -- no additional changes are required in the playbooks.

larsks reviewed (commented):
Here are some initial comments; there will be more as I have additional time :)

tssala23 commented:
@larsks changes were made to this repo as I corrected problems I found whilst creating the cluster. I have removed the machine configs and nodenetworkpolicies now. I intend to add figure out how to add them back but for now we can call this v1.
